### DISCUSSION TITLE
Android activity is incompatible with some libraries

### DISCUSSION THREAD

**Discussion by paulocoutinhox** [axmolengine/axmol#1891] [General]
Hi,

Im trying add RevenueCat to my project on Android side, but `AxmolActivity` inherit from `android.app.Activity` and it don't implement the `ActivityResultCaller`. Because this, some libs that use activity result data don't work.

Issue:
https://github.com/RevenueCat/purchases-android/issues/1682

We can use other in Axmol or need really be `android.app.Activity`?

Thanks.

**Comment by rh101** [comment]
Is there any reason you can't create a sub-class of `AxmolActivity` in your own project, and inherit from `ActivityResultCaller` to implement the interface?

`public class CustomAxmolActivity extends AxmolActivity implements ActivityResultCaller` etc. etc.

Then in your `AndroidManifest.xml` use `CustomAxmolActivity` instead of `org.axmol.app.AppActivity` as the main activity.

For example:
```
<application
...
        <activity
            android:name="com.CustomAxmolActivity"

...
```

As far as I know, that is how it's supposed to be done, and I even implement something similar in my own projects.  There is nothing to change in the Axmol source code.

**Reply by paulocoutinhox** [reply]
This is all implemented by Android in `ComponentActivity`.

```
public class ComponentActivity extends androidx.core.app.ComponentActivity implements
        ContextAware,
        LifecycleOwner,
        ViewModelStoreOwner,
        HasDefaultViewModelProviderFactory,
        SavedStateRegistryOwner,
        OnBackPressedDispatcherOwner,
        ActivityResultRegistryOwner,
        ActivityResultCaller,
        OnConfigurationChangedProvider,
        OnTrimMemoryProvider,
        OnNewIntentProvider,
        OnMultiWindowModeChangedProvider,
        OnPictureInPictureModeChangedProvider,
        MenuHost {}
```

I change the `AxmolActivity` to `AppCompatActivity` and everything still works.

I can make a PR if everybody agree.

**Reply by rh101** [reply]
> This is all implemented by Android in `ComponentActivity`.
> 
That is not relevant, as you stated in your initial post that the only `ActivityResultCaller` needs to be implemented, which should be handled in your own project.

> I change the `AxmolActivity` to `AppCompatActivity` and everything still works.
> 
> I can make a PR if everybody agree.

This is a requirement that is specific to your own project, so should be handled in the project, not in the template or Axmol source code.  If there is actual benefit to all developers in moving to `ComponentActivity`, then present your reasoning clearly.

**Reply by paulocoutinhox** [reply]
Only need change `AxmolActivity` to `AppCompatActivity`. Theme in android manifest to `android:theme="@style/Theme.AppCompat.Light.NoActionBar"`. Add dependency to build.gradle `implementation 'androidx.appcompat:appcompat:1.6.1'`. I think that all real apps need something that will use it, not only mine, no? Im only discussing about.

**Reply by rh101** [reply]
> I think that all real apps need something that will use it, not only mine, no?

I haven't personally experienced that to be the case in any "real app" developed with Axmol or Cocso2d-x, so it may be a little presumptuous to assume every app needs it.

As I mentioned earlier, if you think there is a need to change the base class of AxmolActivity, then please be very clear in presenting your reasons why this is the case.

**Reply by paulocoutinhox** [reply]
There are some advantages, compatibilities and new things support: https://www.quora.com/Whats-the-difference-between-AppCompatActivity-and-normal-Activity-And-what-should-I-use

It will inherit from the same class in the end, but the advantage is that it already implements all the day-to-day things we need.

Is better we change in Axmol code than each one need modify their copy of Axmol to have access to it.

**Reply by paulocoutinhox** [reply]
Some examples to show that AppCompatActivity is more compatible with frameworks that we use :

- Firebase that is inside most app: https://firebase.google.com/docs/auth/android/google-signin?hl=en#authenticate_with_firebase
- RevenueCat that is most payment gateway used: https://www.revenuecat.com/docs/tools/paywalls/displaying-paywalls#how-to-display-a-fullscreen-paywall-in-your-app-1
- Jetpack - Jetpack - Google's Jetpack suite of libraries, which includes components like Lifecycle, ViewModel, LiveData, Navigation, and many others, often requires extending AppCompatActivity to properly implement many of its features.
- Fragments - While you can use Fragment with any activity that is an instance of Activity, many methods and integrations in Fragment perform better or require an AppCompatActivity for full access to backward compatibility and integration with other AndroidX libraries.
- Material Components - The library of material design components for Android, part of AndroidX, also benefits from using AppCompatActivity to ensure all user interface components behave consistently across different versions of Android.
- Data Binding and View Binding - These libraries facilitate interactions with UI components and are often used alongside AppCompatActivity for more effective integration and use of features like LiveData to automatically update the UI in response to data changes.
- Toolbar and ActionBar - To use Toolbar as an app's action bar, you need to extend AppCompatActivity, which allows easy and compatible handling of the action bar across various Android versions.

And others.

**Comment by paulocoutinhox** [comment]
Can i make a PR to add this feature and use AppCompatActivity?
